### ISSUE TITLE
Improve compilation speed for generated codes with automatically marginalized MCMC

### ISSUE THREAD

**Issue by lll6924** [google/jax#17335]
Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

Hi JAX team, this is Jinlin Lai, an author of the ICML paper [Automatically Marginalized MCMC in Probabilistic Programming](https://arxiv.org/pdf/2302.00564.pdf). Thank you for inviting me to open this issue at AABI! 

When doing this work, we have observed excessive compilation time for the gradient function of the log density function of certain models, especially when a node with hundreds of children is marginalized. As you could imagine very complicated codes can be generated by Algorithm 1 and Algorithm 2 in our paper. A small model provided in Appendix G is

$$
x\sim \mathcal{N}(0,1), \log\sigma\sim\mathcal{N}(0,1), y_i\sim\mathcal{N}(x,\sigma^2),
$$

where $y_1,...y_N$ are observed. After marginalizing $x$, codes of $\mathcal{O}(N)$ length are generated, yet the compilation time grows super-linear with $N$. This can also be reproduced with manual marginalization. The following codes are the marginalized model by manually executing Algorithm 1 and Algorithm 2.

```python
import jax.numpy as jnp
from jax import grad, jit
import numpyro.distributions as dist

def model(x, log_s, y):
    N = len(y)
    ans = dist.Normal().log_prob(x) + dist.Normal().log_prob(log_s)
    s = jnp.exp(log_s)

    x_m = 0.
    x_s = 1.

    for i in range(N):
        k = x_s * x_s / (x_s * x_s + s * s)
        y_m = x_m
        y_s = jnp.sqrt(x_s * x_s + s * s)
        ans += dist.Normal(y_m, y_s).log_prob(y[i])
        x_m = x_m + k * (y[i] - y_m)
        x_s = jnp.sqrt(1 - k) * x_s

    return ans

f_jit = jit(model)
g = grad(model, argnums=(0, 1, 2))
g_jit = jit(g)

print(f_jit(0.,0.,jnp.zeros(500)))
print(g_jit(0.,0.,jnp.zeros(500)))
```

When running the codes, `f_jit` takes seconds, but `g_jit` is super slow. With additional experiments it can be found that the lines of Jaxprs in `g` do not grow significantly over `f` (and still stay linear with respect to $N$), so it could be more related to the structure of the computation graph. I am wondering what can be done in the compiler's side to improve the `jit` performance of functions like this. 

Note that the codes are written to simulate generated codes. In the generated codes the for loop is completely unrolled. With manual inspection it is possible to optimize the above codes with cond, but it is beyond my knowledge to automatically generate cond given a graphical model representation.
